compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -6188,19 +6188,74 @@ TR::Register *OMR::X86::TreeEvaluator::mRegStoreEvaluator(TR::Node *node, TR::Co
     return TR::TreeEvaluator::unImpOpEvaluator(node, cg);
 }
 
+TR::Register *OMR::X86::TreeEvaluator::binaryMaskEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+{
+    TR::Node *lhsNode = node->getFirstChild();
+    TR::Node *rhsNode = node->getSecondChild();
+    TR::Register *lhsReg = cg->evaluate(lhsNode);
+    TR::Register *rhsReg = cg->evaluate(rhsNode);
+    TR::Register *resultReg = cg->allocateRegister(lhsReg->getKind());
+    TR::InstOpCode vectorOpcode = TR::InstOpCode::bad;
+    TR::InstOpCode maskOpcode = TR::InstOpCode::bad;
+    bool avx512BW = cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512BW);
+    int32_t numLanes = node->getDataType().getVectorNumLanes();
+
+    switch (node->getOpCode().getVectorOperation()) {
+        case TR::mand:
+            vectorOpcode = TR::InstOpCode::PANDRegReg;
+            maskOpcode = avx512BW ? TR::InstOpCode::KANDQRegRegReg : TR::InstOpCode::KANDWRegRegReg;
+            break;
+        case TR::mor:
+            vectorOpcode = TR::InstOpCode::PORRegReg;
+            maskOpcode = avx512BW ? TR::InstOpCode::KORQRegRegReg : TR::InstOpCode::KORWRegRegReg;
+            break;
+        case TR::mxor:
+            vectorOpcode = TR::InstOpCode::PXORRegReg;
+            maskOpcode = avx512BW ? TR::InstOpCode::KXORQRegRegReg : TR::InstOpCode::KXORWRegRegReg;
+            break;
+        default:
+            TR_ASSERT_FATAL(false, "Unsupported operation in binaryMaskEvaluator");
+            break;
+    }
+
+    if (resultReg->getKind() == TR_VMR) {
+        TR_ASSERT_FATAL(numLanes <= 16 || avx512BW,
+            "binaryMaskEvaluator: Native mask operations on more than 16 lanes requires AVX-512BW");
```

Comment:
If you think there's any chance this assertion could be hit in the future, it would probably help to include more information - maybe use `TR_ASSERT_FATAL_WITH_NODE` and report the values of `numValues` and `avx512BW`.